ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore/Lockstep.hs

```diff
@@ -0,0 +1,811 @@
+        , wosl < at sl
+        , 1 <= Mock.valuesLength vs
+        -> s {readAfterWrite = True}
+      _ -> s
+
+    updateRangeReadAfterWrite :: Stats ks vs d -> Stats ks vs d
+    updateRangeReadAfterWrite s = case (action, result) of
+      (BSVHRangeRead h _, MEither (Right (MValues vs)))
+        | h' <- getHandle $ lookUp h
+        , Just wosl <- Map.lookup h' handleSlots
+        , Just (sl, _) <- Map.lookupMax writeSlots
+        , wosl < at sl
+        , 1 <= Mock.valuesLength vs
+        -> s {rangeReadAfterWrite = True}
+      _ -> s
+
+data TagAction =
+    TBSInitFromValues
+  | TBSInitFromCopy
+  | TBSClose
+  | TBSCopy
+  | TBSValueHandle
+  | TBSWrite
+  | TBSVHClose
+  | TBSVHRangeRead
+  | TBSVHRead
+  | TBSVHAtSlot
+  | TBSVHStat
+  deriving (Show, Eq, Ord, Bounded, Enum)
+
+-- | Identify actions by their constructor.
+tAction :: LockstepAction (BackingStoreState ks vs d) a -> TagAction
+tAction = \case
+  BSInitFromValues _ _ -> TBSInitFromValues
+  BSInitFromCopy _     -> TBSInitFromCopy
+  BSClose              -> TBSClose
+  BSCopy _             -> TBSCopy
+  BSValueHandle        -> TBSValueHandle
+  BSWrite _ _          -> TBSWrite
+  BSVHClose _          -> TBSVHClose
+  BSVHRangeRead _ _    -> TBSVHRangeRead
+  BSVHRead _ _         -> TBSVHRead
+  BSVHAtSlot _         -> TBSVHAtSlot
+  BSVHStat _           -> TBSVHStat
+
+data Tag =
+    -- | A value handle is created before a write, and read after the write. The
+    -- write should not affect the result of the read.
+    ReadAfterWrite
+    -- | A value handle is created before a write, and read after the write. The
+    -- write should not affect the result of the read.
+  | RangeReadAfterWrite
+  | ErrorBecauseBackingStoreIsClosed TagAction
+  | ErrorBecauseBackingStoreValueHandleIsClosed TagAction
+  deriving (Show)
+
+tagBSAction ::
+     Stats ks vs d
+  -> LockstepAction (BackingStoreState ks vs d) a
+  -> BSVal ks vs d a
+  -> [Tag]
+tagBSAction stats action result =
+    globalTags ++ case (action, result) of
+      (_, MEither (Left (MErr ErrBackingStoreClosed))) ->
+        [ErrorBecauseBackingStoreIsClosed (tAction action)]
+      (_, MEither (Left (MErr ErrBackingStoreValueHandleClosed))) ->
+        [ErrorBecauseBackingStoreValueHandleIsClosed (tAction action)]
+      _ -> []
+  where
+    Stats{readAfterWrite, rangeReadAfterWrite} = stats
+
+    globalTags = mconcat [
+        [ ReadAfterWrite
+        | readAfterWrite
+        ]
+      , [ RangeReadAfterWrite
+        | rangeReadAfterWrite
+        ]
+      ]
+
+{-------------------------------------------------------------------------------
+  Errors
+-------------------------------------------------------------------------------}
+
+mkHandler ::
+     (IOLike m, Exception e)
+  => (e -> Maybe Err)
+  -> Handler m (Either Err a)
+mkHandler fhandler = Handler $
+  \e -> maybe (throwIO e) (return . Left) (fhandler e)
+
+-- | Map LMDB errors to mock errors.
```

Comment:
How should I interpret `Nothing` here? Errors that will crash the test suite, maybe?

Reply:
If `Nothing`, then indeed it will lead to the test suite crashing.